server/mediaserver.go

```diff
@@ -1037,6 +1038,85 @@ func getPlaylistsFromStore(ctx context.Context, sess drivers.OSSession, manifest
 	return filesMap, jsonFiles, latestPlaylistTime, nil
 }
 
+func (s *LivepeerServer) streamMP4(w http.ResponseWriter, r *http.Request, jpl *core.JsonPlaylist, manifestID, track string) {
+	w.Header().Set("Access-Control-Allow-Origin", "*")
+	contentType, _ := common.TypeByExtension(".mp4")
+	w.Header().Set("Content-Type", contentType)
+	w.Header().Set("Connection", "keep-alive")
+
+	tc := ffmpeg.NewTranscoder()
+	or, ow, err := os.Pipe()
+	if err != nil {
+		glog.Error(err)
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+	defer func() {
+		tc.StopTranscoder()
+		ow.Close()
+	}()
+	go io.Copy(w, or)
+	oname := fmt.Sprintf("pipe:%d", ow.Fd())
+	out := []ffmpeg.TranscodeOptions{
+		{
+			Oname: oname,
+			VideoEncoder: ffmpeg.ComponentOptions{
+				Name: "copy",
+			},
+			AudioEncoder: ffmpeg.ComponentOptions{
+				Name: "copy",
+			},
+			Profile: ffmpeg.VideoProfile{Format: ffmpeg.FormatNone},
+			Muxer: ffmpeg.ComponentOptions{
+				Name: "mp4",
+				// main option is 'frag_keyframe' which tells ffmpeg to create fragmented MP4 (which we need to be able to stream generatd file)
+				// other options is not mandatory but they will slightly improve generated MP4 file
+				Opts: map[string]string{"movflags": "frag_keyframe+negative_cts_offsets+omit_tfhd_offset+disable_chpl+default_base_moof"},
+			},
+		},
+	}
+	for _, seg := range jpl.Segments[track] {
+		if seg.GetDiscontinuity() {
+			tc.Discontinuity()
+		}
+
+		ir, iw, err := os.Pipe()
+		if err != nil {
+			glog.Error(err)
+			w.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+		fname := fmt.Sprintf("pipe:%d", ir.Fd())
+
+		in := &ffmpeg.TranscodeOptionsIn{Fname: fname, Transmuxing: true}
+		go func() {
+			glog.V(common.VERBOSE).Infof("Adding mid=%s track=%s uri=%s to mp4", manifestID, track, seg.URI)
+			resp, err := http.Get(seg.URI)
+			if err != nil {
+				glog.Errorf("Error getting HTTP uri=%s err=%v", seg.URI, err)
+				return
+			}
+			defer resp.Body.Close()
+			if resp.StatusCode != 200 {
+				glog.Errorf("Non-200 response for status=%v uri=%s request=%s", resp.Status, seg.URI, r.URL.String())
+				return
+			}
+			_, err = io.Copy(iw, resp.Body)
+			if err != nil {
+				glog.Errorf("Error transmuxing to mp4 request=%s uri=%s err=%v", r.URL.String(), seg.URI, err)
+			}
+			iw.Close()
```

Comment:
Do we need to make sure Close() is called if we hit an error condition earlier on in this goroutine?

Reply:
fixed in 50daabe9c6c91455a8f63f9913df6aca58e86494